docs/api/web-frame-main.md

```diff
@@ -231,6 +231,29 @@ A `string` representing the [visibility state](https://developer.mozilla.org/en-
 
 See also how the [Page Visibility API](browser-window.md#page-visibility) is affected by other Electron APIs.
 
+#### `frame.unresponsiveDocumentJSCallStack` _Readonly_ _Experimental_
```

Comment:
this is a strange API name for this shape. Does this only give you a call stack if the document is unresponsive? what happens if you access it while the document is _not_ unresponsive? If it's required that the document be unresponsive for this to work, should this instead be passed as part of the unresponsive event? How should a caller know whether it's available or not?

Reply:
In Chromium, when a WebContents becomes unresponsive, it triggers a Mojo IPC to collect the JS call stack via [content::RenderProcessHostImpl::InterruptJavaScriptIsolateAndCollectCallStack()](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/web_contents/web_contents_impl.cc;l=9385-9395;drc=1d694679493c7b2f7b9df00e967b4f8699321093). There's no observer event to indicate when the collection has completed unfortunately. Currently it's read when a crash report is being generated.

If we were to patch Blink or propose a change upstream, a better API might be
```js
webContents.on('unresponsive-js-call-stack-updated', ({ callStack }) => {
  console.log('Updated call stack', callStack);
})
```

In this feature's current state, this will only ever be accessible if the Chrome feature is enabled with the `Document-Policy` header also set on the unresponsive webpage.

Given the draft status and early implementation, I was leaning towards this being an experimental API for now until something cleaner is available. I'd also be ok with just not documenting this API yet until the standard is finalized.

Reply:
Alternative idea: once the call stack is received, emit `unresponsive` again, now with the call stack details. Again though, this would require a chromium patch.

Reply:
The interface used for capturing the stack https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/public/mojom/call_stack_generator/call_stack_generator.mojom does not specify the requirement of document being in unresponsive state, it seems to be valid to call at any point in time of main thread execution.

Option 1) Could we get a handle to this interface exposed in the public API of `RenderProcessHost`, so that we can listen on the interface response to know when stack collection is compelete. This would make the `WebFrameMain` api more generic.

Option 2) If the feature requirements are met, then `unresponsive` event is emitted with stack details only after the stack collection is complete. This does require changes in upstream when it emits `WebContentsObserver::OnRendererUnresponsive` but takes away the need for a user to call into additional api when they have already signaled that they need stack details.

Reply:
> In this feature's current state, this will only ever be accessible if the Chrome feature is enabled with the Document-Policy header also set on the unresponsive webpage.

It would be nice for this to not be required, though it's probably not worth maintaining an extra patch since it seems fairly easy to work around.

Reply:
> Could we get a handle to this interface exposed in the public API of RenderProcessHost, so that we can listen on the interface response to know when stack collection is compelete.

Once I have more time to focus on this, I'll give this a try.

> If the feature requirements are met, then unresponsive event is emitted with stack details only after the stack collection is complete.

I believe there are cases where it's unsuccessful in capturing the call stack. Would need to measure if that adds additional time to the round-trip IPCs. That would be convenient if it could be included with the event!

> It would be nice for this to not be required, though it's probably not worth maintaining an extra patch since it seems fairly easy to work around.

To clarify, this feature will eventually be enabled by default presumably once the draft and implementation upstream has completed.

-----

btw, I've had luck using this in our application by enabling the necessary feature flags. It sends the call stacks to the reporting server. Would recommend trying that route at least before this feature is available!

Reply:
If we have the APIs to be able to collect a call stack at any time, let’s expose that!

if we can *additionally* hook into logic that chrome uses to collect call stacks when unresponsive, that seems useful too, but we already have logic for detecting unresponsive frames and if we had an API for grabbing a stack then clients could just use it in their event handlers.